web3/gas_strategies/time_based.py

```diff
@@ -0,0 +1,152 @@
+import operator
+import collections
+import math
+
+from cytoolz import (
+    curry,
+)
+
+from cytoolz import (
+    groupby,
+    sliding_window,
+)
+
+from eth_utils import (
+    to_tuple,
+)
+
+
+MinerData = collections.namedtuple('MinerData', ['miner', 'num_blocks', 'min_gas_price'])
+Probability = collections.namedtuple('Probability', ['gas_price', 'prob'])
+
+
+def get_avg_block_time(w3, sample_size):
+    latest = w3.eth.getBlock('latest')
+    oldest = w3.eth.getBlock(max(0, latest['number'] - sample_size))
+    return (latest['timestamp'] - oldest['timestamp']) / sample_size
+
+
+def get_raw_miner_data(w3, sample_size):
+    latest = w3.eth.getBlock('latest', full_transactions=True)
+
+    for transaction in latest['transactions']:
+        yield (latest['miner'], latest['hash'], transaction['gasPrice'])
+
+    block = latest
+
+    for _ in range(sample_size - 1):
+        if block['number'] == 0:
+            break
+        # we intentionally trace backwards using parent hashes rather than
+        # block numbers to make caching the data easier to implement.
+        block = w3.eth.getBlock(block['parentHash'], full_transactions=True)
+        for transaction in block['transactions']:
+            yield (block['miner'], block['hash'], transaction['gasPrice'])
+
+
+def aggregate_miner_data(raw_data):
+    data_by_miner = groupby(1, raw_data)
+
+    for miner, miner_data in data_by_miner.items():
+        _, block_hashes, gas_prices = map(set, zip(*miner_data))
+        yield MinerData(miner, len(set(block_hashes)), min(gas_prices))
+
+
+@to_tuple
+def compute_probabilities(miner_data, wait_blocks, sample_size):
+    """
+    Computes the probabilities that a txn will be accepted at each of the gas
+    prices accepted by the miners.
+    """
+    miner_data_by_price = tuple(sorted(
+        miner_data,
+        key=operator.attrgetter('min_gas_price'),
+        reverse=True,
+    ))
+    for idx in range(len(miner_data_by_price)):
+        min_gas_price = miner_data_by_price[idx].min_gas_price
+        num_blocks_accepting_price = sum(m.num_blocks for m in miner_data_by_price[idx:])
+        inv_prob_per_block = (sample_size - num_blocks_accepting_price) / sample_size
+        probability_accepted = 1 - inv_prob_per_block ** wait_blocks
+        yield Probability(min_gas_price, probability_accepted)
```

Comment:
This looks like an optimistic probability calculation. If a miner mines 20 blocks out of a sample of 100, and includes one transaction at a 1 gwei price, then this treats it as if a 1 gwei transaction is just as likely to get included in all 20 blocks.

Am I reading that right?

Reply:
Yes, I'm open to suggestions for how to account for this.

Also, here's a list of other things I know that aren't currently being accounted for.

- transaction `gas` limit.
- competing transactions in the transaction pool.

I'm sure there are others.  The rabbit hole is deep.

Reply:
> suggestions for how to account for this

Maybe track the lowest gwei per block, ignoring the miner for this calculation? That may be a bad idea for the increase in required data. Frankly, I am still a little pessimistic about the approach of estimating gas price at all, so I suppose I'm in the camp of limit effort and I will just say "it's fine as is."

(The alternative to estimating gas price I am thinking is a daemon that watches your pending transactions and repeatedly bumps the price if they haven't been picked up for a while)

> The rabbit hole is deep.

For real